File: Raven.Client.Lightweight/Document/OAuth/BasicAuthenticator.cs
using System;
using System.IO;
using System.Net;
using System.Threading.Tasks;
using Raven.Abstractions.Connection;
using Raven.Client.Connection;
using Raven.Client.Extensions;
#if SILVERLIGHT
using Raven.Client.Silverlight.Connection;
using System.Net.Browser;
#endif

namespace Raven.Client.Document.OAuth
{
	public class BasicAuthenticator : AbstractAuthenticator
	{
		private readonly string apiKey;
		private readonly HttpJsonRequestFactory jsonRequestFactory;
	
		public BasicAuthenticator(string apiKey, HttpJsonRequestFactory jsonRequestFactory)
		{
			this.apiKey = apiKey;
			this.jsonRequestFactory = jsonRequestFactory;
		}


		public Task<Action<HttpWebRequest>> HandleOAuthResponseAsync(string oauthSource)
		{
			var authRequest = PrepareOAuthRequest(oauthSource);
			return Task<WebResponse>.Factory.FromAsync(authRequest.BeginGetResponse, authRequest.EndGetResponse, null)
				.AddUrlIfFaulting(authRequest.RequestUri)
				.ConvertSecurityExceptionToServerNotFound()
				.ContinueWith(task =>
				{
#if !SILVERLIGHT
					using (var stream = task.Result.GetResponseStreamWithHttpDecompression())
#else
					using(var stream = task.Result.GetResponseStream())
#endif
					using (var reader = new StreamReader(stream))
					{
						CurrentOauthToken = "Bearer " + reader.ReadToEnd();
						return (Action<HttpWebRequest>) (request => SetHeader(request.Headers, "Authorization", CurrentOauthToken));
					}
				});
		}

#if !SILVERLIGHT
		public Action<HttpWebRequest> HandleOAuthResponse(string oauthSource)
		{
			var authRequest = PrepareOAuthRequest(oauthSource);
			using (var response = authRequest.GetResponse())
			{
				using (var stream = response.GetResponseStreamWithHttpDecompression())
				using (var reader = new StreamReader(stream))
				{
					CurrentOauthToken = "Bearer " + reader.ReadToEnd();
					return request => SetHeader(request.Headers, "Authorization", CurrentOauthToken);
				}
			}
		}
#endif

		private HttpWebRequest PrepareOAuthRequest(string oauthSource)
		{
#if !SILVERLIGHT
			var authRequest = (HttpWebRequest)WebRequest.Create(oauthSource);
			authRequest.Headers["Accept-Encoding"] = "deflate,gzip";
#else
			var authRequest = (HttpWebRequest) WebRequestCreator.ClientHttp.Create(new Uri(oauthSource.NoCache()));
#endif
			authRequest.Headers["grant_type"] = "client_credentials";
			authRequest.Accept = "application/json;charset=UTF-8";

			if (String.IsNullOrEmpty(apiKey) == false)
				SetHeader(authRequest.Headers, "Api-Key", apiKey);

			if (oauthSource.StartsWith("https", StringComparison.InvariantCultureIgnoreCase) == false &&
			   jsonRequestFactory.EnableBasicAuthenticationOverUnsecureHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers == false)
				throw new InvalidOperationException(BasicOAuthOverHttpError);
			return authRequest;
		}

		private const string BasicOAuthOverHttpError = @"Attempting to authenticate using basic security over HTTP would expose user credentials (including the password) in clear text to anyone sniffing the network.
Your OAuth endpoint should be using HTTPS, not HTTP, as the transport mechanism.
You can setup the OAuth endpoint in the RavenDB server settings ('Raven/OAuthTokenServer' configuration value), or setup your own behavior by providing a value for:
	documentStore.Conventions.HandleUnauthorizedResponse
If you are on an internal network or requires this for testing, you can disable this warning by calling:
	documentStore.JsonRequestFactory.EnableBasicAuthenticationOverUnsecureHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers = true;
";
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/abstractions/oauth/BasicAuthenticator.java
package raven.abstractions.oauth;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpRequest;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;

import raven.abstractions.closure.Action1;

public class BasicAuthenticator extends AbstractAuthenticator {
  private final String apiKey;
  private final boolean enableBasicAuthenticationOverUnsecuredHttp;

  public BasicAuthenticator(String apiKey, boolean enableBasicAuthenticationOverUnsecuredHttp) {
    this.apiKey = apiKey;
    this.enableBasicAuthenticationOverUnsecuredHttp = enableBasicAuthenticationOverUnsecuredHttp;
  }

  @Override
  public Action1<HttpRequest> doOAuthRequest(String oauthSource) {
    HttpClient httpClient = new DefaultHttpClient();
    try {

      HttpGet authRequest = prepareOAuthRequest(oauthSource);
      HttpResponse httpResponse = httpClient.execute(authRequest);
      try {
        final String response = IOUtils.toString(httpResponse.getEntity().getContent());
        return new Action1<HttpRequest>() {

          @Override
          public void apply(HttpRequest request) {
            request.setHeader("Authorization", "Bearer " + response);
          }
        };
      } finally {
        EntityUtils.consume(httpResponse.getEntity());
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    } finally {
      httpClient.getConnectionManager().shutdown();
    }
  }

  private HttpGet prepareOAuthRequest(String oauthSource) {
    HttpGet get = new HttpGet();
    get.setHeader("grant_type", "client_credentials");
    get.setHeader("Accept", "application/json;charset=UTF-8");
    if (StringUtils.isNotEmpty(apiKey)) {
      get.setHeader("Api-Key", apiKey);
    }

    if (!oauthSource.startsWith("https") && !enableBasicAuthenticationOverUnsecuredHttp) {
      throw new IllegalStateException(BASICO_AUTH_OVER_HTTP_ERROR);
    }

    return get;
  }

  private final static String BASICO_AUTH_OVER_HTTP_ERROR = "Attempting to authenticate using basic security over HTTP would expose user credentials (including the password) in clear text to anyone sniffing the network." +
      "Your OAuth endpoint should be using HTTPS, not HTTP, as the transport mechanism." +
      "You can setup the OAuth endpoint in the RavenDB server settings ('Raven/OAuthTokenServer' configuration value), or setup your own behavior by providing a value for: " +
      "documentStore.Conventions.HandleUnauthorizedResponse " +
      "If you are on an internal network or requires this for testing, you can disable this warning by calling: " +
      "documentStore.getJsonRequestFactory().enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers = true; ";

}
